### ISSUE TITLE
placeBlock() throw an unexpected error.

### ISSUE THREAD

**Issue by EnderTheCoder** [PrismarineJS/mineflayer#2552]
- [x] The [FAQ](https://github.com/PrismarineJS/mineflayer/blob/master/docs/FAQ.md) doesn't contain a resolution to my issue 

## Versions
 - mineflayer: 4.2.0
 - server: paper 1.18.2
 - node: 17.7.2

## Detailed description of a problem
I'm making a mining bot. The bot is supposed to replace liquid block with dirt when encounter. But when it tried to place the block at the liquid's position, it throws an error: "No block has been placed : the block is still dirt".

## What did you try yet?

I have checked that the function that I used to calc reference block and vector is correct. And actually, it did place the block, but I still got this error.
And I have checked that this placeBlock() was executed only once at the same position by adding a "console.log("place is executed")" before the placeBlock(). The message "place is executed" showed only once.
![2022-03-30 02-25-56 的屏幕截图](https://user-images.githubusercontent.com/47713889/160683113-3636f255-6f56-4892-b4a0-0d73575acf5f.png)
![2022-03-30 02-50-25 的屏幕截图](https://user-images.githubusercontent.com/47713889/160683950-73a64da6-3703-41ea-a4be-a368642c40eb.png)
![2022-03-30 02-50-39 的屏幕截图](https://user-images.githubusercontent.com/47713889/160683964-9b690f71-d1d0-4076-8ce1-840f93883523.png)

I actually have already made a issue at https://github.com/PrismarineJS/mineflayer/issues/2546. But nobody replied.
I thought that was cause by using a air block as reference block, but it wasn't. I think they are the same problem.
## Your current code
```js


    private isBlockReferencable(block, offsetX, offsetY, offsetZ) {
        return this.bot.blockAt(block.position.offset(offsetX, offsetY, offsetZ)).name != "air" &&
            this.bot.blockAt(block.position.offset(offsetX, offsetY, offsetZ)).name != "lava" &&
            this.bot.blockAt(block.position.offset(offsetX, offsetY, offsetZ)).name != "water"
    }

    public findReferencableBlock() {
        let referenceBlock
        let faceVector
        if (this.isBlockReferencable(this.targetBlock, 1, 0, 0)) {
            referenceBlock = this.bot.blockAt(this.targetBlock.position.offset(1, 0, 0))
            faceVector = new Vec3(-1, 0, 0)
        } else if (this.isBlockReferencable(this.targetBlock, -1, 0, 0)) {
            referenceBlock = this.bot.blockAt(this.targetBlock.position.offset(-1, 0, 0))
            faceVector = new Vec3(1, 0, 0)
        } else if (this.isBlockReferencable(this.targetBlock, 0, 1, 0)) {
            referenceBlock = this.bot.blockAt(this.targetBlock.position.offset(0, 1, 0))
            faceVector = new Vec3(0, -1, 0)
        } else if (this.isBlockReferencable(this.targetBlock, 0, -1, 0)) {
            referenceBlock = this.bot.blockAt(this.targetBlock.position.offset(0, -1, 0))
            faceVector = new Vec3(0, 1, 0)
        } else if (this.isBlockReferencable(this.targetBlock, 0, 0, 1)) {
            referenceBlock = this.bot.blockAt(this.targetBlock.position.offset(0, 0, 1))
            faceVector = new Vec3(0, 0, -1)
        } else if (this.isBlockReferencable(this.targetBlock, 0, 0, -1)) {
            referenceBlock = this.bot.blockAt(this.targetBlock.position.offset(0, 0, -1))
            faceVector = new Vec3(0, 0, 1)
        } else {
            referenceBlock = this.bot.blockAt(this.targetBlock.position.offset(1, 0, 0))
            faceVector = new Vec3(-1, 0, 0)
            console.warn("WARNING: No legal reference block found. Use illegal block instead. This might cause error or ban.")
        }

        return {referenceBlock: referenceBlock, faceVector: faceVector}
    }

    public async fillLiquid() {

        let reference = this.findReferencableBlock()
        let inventory = new BotInventory(this.bot)

        // for (let i = 0; i < this.garbageList.length; i++) {
        //     if (inventory.countItem(this.garbageList[i]) > 0) {
        //         await this.bot.equip(this.mcData.itemsByName[this.garbageList[i]], "hand")
        //         break
        //     }
        // }
        try {
            console.log("place is executed")

            await this.bot.placeBlock(reference.referenceBlock, reference.faceVector)
        } catch (e) {
            console.warn(e)
        }
    }

```

## Expected behavior
Fix this unexpected error throwing.

**Comment by IceTank** [comment]
This might have something to do with the other issue you posted about block placement. If this is specific to 1.18.2 then this is probably something that changed from 1.17 to 1.18. If you want to look into it then the issue seams to be replacing blocks that are replaceable and how mineflayer handles the server response for doing so.

**Comment by creator1creeper1** [comment]
I'm not sure which issue is the right one to post in since they basically cover the same bug.

I just wanted to say that I've just been bitten by this - of course it's not correct to place blocks in mid-air, but the current error message is pretty misleading. The error even occurs if the block *is* being placed (even if the server really shouldn't allow it). So it would be useful to either somehow allow this case or give a better error message.

**Comment by IceTank** [comment]
> I just wanted to say that I've just been bitten by this - of course it's not correct to place blocks in mid-air, but the current error message is pretty misleading. The error even occurs if the block is being placed (even if the server really shouldn't allow it). So it would be useful to either somehow allow this case or give a better error message.

But the server response is exactly what the error says. It is just that the block from the bot's point off view is placed in the wrong position when placing it mid air with the testing I did. There is also some weird block updating going on that might also effect how the error is displayed. The issue is that placing blocks in mid air (or maybe replacing blocks ?) can be undefined behavior causing an error. This just happened to just work prior to 1.18 but now that the minecraft server response is different it broke it. Mineflayer is not intended to handle none vanilla blocks placement as mineflayer aims to mirror the vanilla client. You can make a better error message or fix placeBlock to handle those specific edge cases when placing blocks mid air. 

If there is another issue with replacing blocks then make a new issue about it.

**Comment by EnderTheCoder** [comment]
But this is an issue with replacing blocks, not blocks in the air. This works with vanilla client, and so that mineflayer is also supposed to be able to do this.

**Comment by EnderTheCoder** [comment]
The problem is simple, it is able to replace blocks, but throw an error with this legal movement.

**Comment by IceTank** [comment]
Hm ok I see I was confused by another issue about replacing blocks. Yeah so someone has to check how exactly vanilla works and fix mineflayer then.

**Comment by EnderTheCoder** [comment]
true
